public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.test.ts

```diff
@@ -0,0 +1,197 @@
+    throw new Error('multiple cursors');
+  }
+
+  const result = getSituation(text, pos);
+
+  if (expectedSituation === null) {
+    expect(result).toStrictEqual(null);
+  } else {
+    expect(result).toMatchObject(expectedSituation);
+  }
+}
+
+describe('situation', () => {
+  it('handles things', () => {
+    assertSituation('^', {
+      type: 'EMPTY',
+    });
+
+    assertSituation('s^', {
+      type: 'AT_ROOT',
+    });
+
+    assertSituation('{level="info"} ^', {
+      type: 'AFTER_SELECTOR',
+      afterPipe: false,
+      labels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    // should not trigger AFTER_SELECTOR before the selector
+    assertSituation('^ {level="info"}', null);
+
+    // check for an error we had during the implementation
+    assertSituation('{level="info" ^', null);
+
+    assertSituation('{level="info"} | json ^', {
+      type: 'AFTER_SELECTOR',
+      afterPipe: false,
+      labels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    assertSituation('{level="info"} | json | ^', {
+      type: 'AFTER_SELECTOR',
+      afterPipe: true,
+      labels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    assertSituation('count_over_time({level="info"}^[10s])', {
+      type: 'AFTER_SELECTOR',
+      afterPipe: false,
+      labels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    // should not trigger AFTER_SELECTOR before the selector
+    assertSituation('count_over_time(^{level="info"}[10s])', null);
+
+    // should work even when the query is half-complete
+    assertSituation('count_over_time({level="info"}^)', {
+      type: 'AFTER_SELECTOR',
+      afterPipe: false,
+      labels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    /*
+    Currently failing, reason unknown
+    assertSituation('sum(^)', {
+      type: 'IN_AGGREGATION',
+    });*/
+  });
+
+  it('handles label names', () => {
+    assertSituation('{^}', {
+      type: 'IN_LABEL_SELECTOR_NO_LABEL_NAME',
+      otherLabels: [],
+    });
+
+    assertSituation('sum(count_over_time({level="info"})) by (^)', {
+      type: 'IN_GROUPING',
+      otherLabels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    assertSituation('sum by (^) (count_over_time({level="info"}))', {
+      type: 'IN_GROUPING',
+      otherLabels: [{ name: 'level', value: 'info', op: '=' }],
+    });
+
+    assertSituation('{one="val1",two!="val2",three=~"val3",four!~"val4",^}', {
+      type: 'IN_LABEL_SELECTOR_NO_LABEL_NAME',
+      otherLabels: [
+        { name: 'one', value: 'val1', op: '=' },
+        { name: 'two', value: 'val2', op: '!=' },
+        { name: 'three', value: 'val3', op: '=~' },
+        { name: 'four', value: 'val4', op: '!~' },
+      ],
+    });
+
+    assertSituation('{one="val1",^}', {
+      type: 'IN_LABEL_SELECTOR_NO_LABEL_NAME',
+      otherLabels: [{ name: 'one', value: 'val1', op: '=' }],
+    });
+
+    // single-quoted currently failing in lezer-logql
```

Comment:
@gabor @ivanahuckova I had to remove the test case for single quotes (`{one='val\'1',^}`) from the test, because currently `lezer-logql` process the query as errors:

![imagen](https://user-images.githubusercontent.com/1069378/191993318-d1d9eddd-09ab-4ac3-b359-0d201cb51749.png)

Unline double quotes or ticks:

![imagen](https://user-images.githubusercontent.com/1069378/191993245-966345fc-281d-4860-85ae-d674e3d223eb.png)

Not sure if this is expected or not. In the hackathon PR there was a test case for single quotes.

Reply:
@matyax , that may be a mistake in the hackathon-code. i have just tried to run a query like this in Loki `{place='on\'e'}`, and it said 
"parse error at line 1, col 8: invalid char literal"

so, maybe `\'` does not exist :) . pinging @ivanahuckova for a definitive answer.